src/storage/src/source/generator/key_value.rs

```diff
@@ -0,0 +1,584 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+use std::convert::Infallible;
+
+use differential_dataflow::{AsCollection, Collection};
+use futures::stream::StreamExt;
+use mz_ore::cast::CastFrom;
+use mz_repr::{Datum, Diff, Row};
+use mz_storage_types::sources::load_generator::KeyValueLoadGenerator;
+use mz_storage_types::sources::{MzOffset, SourceTimestamp};
+use mz_timely_util::builder_async::{OperatorBuilder as AsyncOperatorBuilder, PressOnDropButton};
+use rand::rngs::StdRng;
+use rand::{RngCore, SeedableRng};
+use timely::dataflow::operators::{Concat, ToStream};
+use timely::dataflow::{Scope, Stream};
+use timely::progress::Antichain;
+
+use crate::healthcheck::{HealthStatusMessage, HealthStatusUpdate, StatusNamespace};
+use crate::source::types::ProgressStatisticsUpdate;
+use crate::source::{RawSourceCreationConfig, SourceMessage, SourceReaderError};
+
+pub fn render<G: Scope<Timestamp = MzOffset>>(
+    key_value: KeyValueLoadGenerator,
+    scope: &mut G,
+    config: RawSourceCreationConfig,
+    resume_uppers: impl futures::Stream<Item = Antichain<MzOffset>> + 'static,
+    start_signal: impl std::future::Future<Output = ()> + 'static,
+) -> (
+    Collection<G, (usize, Result<SourceMessage, SourceReaderError>), Diff>,
+    Option<Stream<G, Infallible>>,
+    Stream<G, HealthStatusMessage>,
+    Stream<G, ProgressStatisticsUpdate>,
+    Vec<PressOnDropButton>,
+) {
+    let (steady_state_stats_stream, stats_button) =
+        render_statistics_operator(scope, config.clone(), resume_uppers);
+
+    let mut builder = AsyncOperatorBuilder::new(config.name.clone(), scope.clone());
+
+    let (mut data_output, stream) = builder.new_output();
+    let (_progress_output, progress_stream) = builder.new_output();
+    let (mut stats_output, stats_stream) = builder.new_output();
+
+    let button = builder.build(move |caps| async move {
+        let [mut cap, mut progress_cap, stats_cap]: [_; 3] = caps.try_into().unwrap();
+
+        let resume_upper = Antichain::from_iter(
+            config.source_resume_uppers[&config.id]
+                .iter()
+                .map(MzOffset::decode_row),
+        );
+
+        let Some(resume_offset) = resume_upper.into_option() else {
+            return;
+        };
+
+        cap.downgrade(&resume_offset);
+        progress_cap.downgrade(&resume_offset);
+        start_signal.await;
+
+        let snapshotting = resume_offset.offset == 0;
+
+        let mut local_partitions: BTreeMap<_, _> = (0..key_value.partitions)
```

Comment:
Why is this a map? The keys are never read

Reply:
No real reason, other than the fact that I don't need to reason about the `Ord` impl of the generator structs